core/src/main/java/org/fao/geonet/resources/CMISResources.java

```diff
@@ -0,0 +1,491 @@
+                String parentKey = key.substring(0, lastFolderDelimiterKeyIndex);
+
+                Map<String, Object> properties = new HashMap<String, Object>();
+                properties.put(PropertyIds.OBJECT_TYPE_ID, "cmis:document");
+                properties.put(PropertyIds.NAME, filenameKey);
+
+                InputStream stream = Files.newInputStream(path);
+                ContentStream contentStream = CMISConfiguration.getClient().getObjectFactory().createContentStream(key, Files.size(path), Files.probeContentType(path), stream);
+
+                if (CMISConfiguration.getClient().existsPath(key)) {
+                    try {
+                        // If the document is found then we are updating the existing document.
+                        Document doc = (Document) CMISConfiguration.getClient().getObjectByPath(key, oc);
+                        doc.updateProperties(properties, true);
+                        doc.setContentStream(contentStream, true, true);
+
+                        Log.info(Geonet.RESOURCES,
+                                String.format("Updated resource '%s'. Current version '%s'.", key, doc.getVersionLabel()));
+                    } catch (CmisPermissionDeniedException ex) {
+                        Log.warning(Geonet.RESOURCES, String.format(
+                                "No permissions to update resource '%s'.", key));
+                        throw new NotAllowedException(String.format(
+                                "No permissions to update resource '%s'.", key));
+                    }
+                } else {
+                    // Get parent folder.
+                    try {
+                        Folder parentFolder;
+                        try {
+                            parentFolder = (Folder) CMISConfiguration.getClient().getObjectByPath(parentKey, oc);
+                        } catch (CmisObjectNotFoundException e) {
+                            // Create parent folder if it does not exists.
+                            ObjectId objectId = CMISConfiguration.getClient().createPath(parentKey, "cmis:folder");
+                            parentFolder = (Folder) CMISConfiguration.getClient().getObject(objectId, oc);
+                        }
+
+                        Document doc = parentFolder.createDocument(properties, contentStream, VersioningState.MAJOR);
+
+                        Log.info(Geonet.RESOURCES,
+                                String.format("Added resource '%s'.", doc.getPaths().get(0)));
+                    } catch (CmisPermissionDeniedException ex) {
+                        Log.warning(Geonet.RESOURCES, String.format(
+                                "No permissions to add resource '%s'.", key));
+                        throw new NotAllowedException(String.format(
+                                "No permissions to add resource '%s'.", key));
+                    }
+                }
+            }
+            java.nio.file.Files.delete(path);
```

Comment:
Should be added in `try / finally` block?

Reply:
It was taken from S3 implementation which did not hava a try catch finally block to handle the errors.
I added it